### PR TITLE
basic support to test against official JSON-Schema-Test-Suite

### PR DESCRIPTION
**Pull request by StarpTech**
This PR will add basic support to test against the official https://github.com/json-schema-org/JSON-Schema-Test-Suite. It contains only tests for the `required` feature.

### CODE REVIEW

**Comment by StarpTech** [comment]
It should be added incrementally or what do you mean?

**Comment by mcollina** [comment]
Would it be hard to just add it to this PR?

**Comment by StarpTech** [comment]
The `required` test-case or all JSON files?

**Comment by StarpTech** [comment]
I added draft 6,7 of the `required` test-case.

**Comment by StarpTech** [comment]
@mcollina one important note: I had to add explicitly the `type: 'object'` property to all schemas otherwise `fast-json-stringify` will throw `undefined unsupported` 😥 acccording to the spec it should fallback to type object when `properties` is used.

**Comment by StarpTech** [comment]
Do we really need to support Node 4 ?

**Comment by mcollina** [comment]
> @mcollina one important note: I had to add explicitly the type: 'object' property to all schemas otherwise fast-json-stringify will throw undefined unsupported 😥 acccording to the spec it should fallback to type object when properties is used.

We should actually behave properly and assume the same, it shouldn't be hard to patch.

> Do we really need to support Node 4 ?

We can bump the major without much problems if it's just for the support.

**Comment by StarpTech** [comment]
> We should actually behave properly and assume the same, it shouldn't be hard to patch.

I think it is very hard because we have to handle the `any` type-case which will make optimizations very complicated. In my opinion it's fine to set some prerequisites in order to build an optimized stringify function this includes to always add the type in your schemas.

e.g 

schema
```json
{
    "properties": {
        "foo": {},
        "bar": {}
    },
    "required": ["foo"]
}
```
this is valid:
```json
{"foo": 1}
```

https://github.com/json-schema/json-schema/issues/172

**Comment by StarpTech** [comment]
> We can bump the major without much problems if it's just for the support.

Great, I will remove node 4 from the matrix.

**Comment by StarpTech** [comment]
Please check again I infer type based on the validation keywords.

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 language: node_js
 sudo: false
 node_js:
-- '4'
 - '6'
 - '8'
 - '9'
```
Comment (mcollina) [inline .travis.yml]: can you drop 9 as well?

File: index.js [hunk 2]
```diff
@@ -114,6 +118,59 @@ function build (schema, options) {
...
+  'pattern'
+]
+
+var numberKeywords = [
+  'multipleOf',
+  'maximum',
+  'exclusiveMaximum',
+  'minimum',
+  'exclusiveMinimum'
+]
```
Comment (mcollina) [inline index.js]: I would just use `const` for these.

**Pull request merged**